core/src/main/java/org/mule/runtime/core/internal/streaming/bytes/PoolingByteBufferManager.java

```diff
@@ -0,0 +1,145 @@
+import static org.apache.commons.pool.impl.GenericObjectPool.WHEN_EXHAUSTED_GROW;
+import static org.mule.runtime.api.i18n.I18nMessageFactory.createStaticMessage;
+import static org.slf4j.LoggerFactory.getLogger;
+import org.mule.runtime.api.exception.MuleRuntimeException;
+import org.mule.runtime.api.lifecycle.Disposable;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import com.google.common.cache.RemovalListener;
+
+import java.nio.ByteBuffer;
+
+import org.apache.commons.pool.ObjectPool;
+import org.apache.commons.pool.PoolableObjectFactory;
+import org.apache.commons.pool.impl.GenericObjectPool;
+import org.slf4j.Logger;
+
+/**
+ * {@link ByteBufferManager} implementation which pools instances for better performance.
+ * <p>
+ * Buffers are kept in separate pools depending on their capacity.
+ * <p>
+ * Idle buffers and capacity pools are automatically expired.
+ *
+ * @since 4.0
+ */
+public class PoolingByteBufferManager implements ByteBufferManager, Disposable {
+
+  private static final Logger LOGGER = getLogger(PoolingByteBufferManager.class);
+
+  private final LoadingCache<Integer, ObjectPool<ByteBuffer>> pools = CacheBuilder.newBuilder()
+      .expireAfterAccess(1, MINUTES)
+      .removalListener((RemovalListener<Integer, ObjectPool<ByteBuffer>>) notification -> {
+        try {
+          notification.getValue().close();
+        } catch (Exception e) {
+          if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("Found exception trying to dispose buffer pool for capacity " + notification.getKey(), e);
+          }
+        }
+      }).build(new CacheLoader<Integer, ObjectPool<ByteBuffer>>() {
+
+        @Override
+        public ObjectPool<ByteBuffer> load(Integer capacity) throws Exception {
+          return createPool(capacity);
+        }
+      });
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public ByteBuffer allocate(int capacity) {
+    try {
+      return pools.getUnchecked(capacity).borrowObject();
+    } catch (Exception e) {
+      throw new MuleRuntimeException(createStaticMessage("Could not allocate byte buffer"), e);
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  @Override
+  public void deallocate(ByteBuffer byteBuffer) {
+    int capacity = byteBuffer.capacity();
+    ObjectPool<ByteBuffer> pool = pools.getIfPresent(capacity);
+    if (pool != null) {
+      try {
+        pool.returnObject(byteBuffer);
+      } catch (Exception e) {
+        if (LOGGER.isDebugEnabled()) {
+          LOGGER.debug("Could not deallocate buffer of capacity " + capacity, e);
+        }
+      }
+    }
+  }
+
+  @Override
+  public void dispose() {
+    pools.invalidateAll();
+  }
+
+  private ObjectPool<ByteBuffer> createPool(int capacity) {
+    GenericObjectPool.Config config = new GenericObjectPool.Config();
+    config.maxIdle = DEFAULT_MAX_IDLE;
```

Comment:
What's the purpose of doing that? This class is not API and serves one and only one purpose. If I make it configurable, there will be nobody configuring it. There's no point in adding that until a use case appears